### ISSUE TITLE
Read APIError response body

### ISSUE THREAD

**Issue by mxinden** [go-openapi/runtime#121]
I make a request via a client generated based on a swagger spec generated via the goswagger.io project with an invalid json property. The back-end returns a HTTP 422 with details in the response body. Reading the response body is crucial to debug the problem.

`go-openapi/runtime/client/runtime.go.Submit()` calls `readResponse.ReadResponse` which returns an `APIError` with the response body embedded:

https://github.com/go-openapi/runtime/blob/231d7876b7019dbcbfc97a7ba764379497b67c1d/client/runtime.go#L422

Problem is, that this embedded response body is closed via a `defer` in `go-openapi/runtime/client/runtime.go.Submit()`:

https://github.com/go-openapi/runtime/blob/231d7876b7019dbcbfc97a7ba764379497b67c1d/client/runtime.go#L395

Thereby casting the error and reading the response body later on fails with `http: read on closed response body`:

```golang
apiError, ok := err.(*runtime.APIError)
if !ok {
	// handle
}

response, ok := apiError.Response.(runtime.ClientResponse)
if !ok {
	// handle
}

body, err := ioutil.ReadAll(response.Body())
if err != nil {
	// fails with "http: read on closed response body"
}

fmt.Println(string(body))
```

How can I access the APIError response body? Might it be possible to embed the body as a string inside the APIError? Let me know if you need more info.

**Comment by mxinden** [comment]
@casualjim any thoughts on this?

**Comment by HexHacks** [comment]
I'm having problem with this also. As an outsider of this project, it seems like the `func (*APIError) Error()` should handle this similarly to how @mxinden handles the error (with some fix to the stated error). To me, the error provided by this function is not sufficient.

**Comment by casualjim** [comment]
you have to provide the error datastructure in the definitions, so that it gets picked up
or put a schema of `type: object` for the response body schema 

the mixin tool lets you edit a spec by deep merging another one, in case the spec isn't under your control.

**Comment by HexHacks** [comment]
Ignore my ranting, my problem is actually 'just' that of @mxinden . If I just print the error by the APIError's own error function, I get this:
`unknown error (status 500): {resp:0xc0000ae3f0}`

When I @mxinden 's approach to explicitly read the body of the `ClientResponse` I get `http: read on closed response body`

**Comment by fjgmmg** [comment]
I was running into the same problem.  Interestingly, when then DEBUG option is on, the response body is cloned and so you don't have the closed response body issue.  So, a really simple fix would be to change the code (https://github.com/go-openapi/runtime/blob/231d7876b7019dbcbfc97a7ba764379497b67c1d/client/runtime.go#L395)  from: 

	defer res.Body.Close()

	if r.Debug {
		b, err2 := httputil.DumpResponse(res, true)
		if err2 != nil {
			return nil, err2
		}
		r.logger.Debugf("%s\n", string(b))
	}

to this:
	defer res.Body.Close()

	b, err2 := httputil.DumpResponse(res, true)
	if err2 != nil {
		return nil, err2
	}
	if r.Debug {
		r.logger.Debugf("%s\n", string(b))
	}

**Comment by deeptibaghel** [comment]
I am facing the same issue, reading the response body is a must in my application. Is there any solution available ?

**Comment by fredbi** [comment]
@deeptibaghel  There is no problem reading a response body, including error responses.
You just have to define your response with a non empty schema.

Like this:
```yaml
responses:
  badRequest:
    description: client error in request
    schema:
      type: object
      properties:
        errorMessage:
          type: string
```

If you expect a generic object, do this like:
```yaml
responses:
  badRequest:
    description: client error in request
    schema:
      type: object
      additionalProperties: true
```

**Comment by deeptibaghel** [comment]
Thanks @fredbi , so my swagger yaml file should have it and then the client will generate methods to read it right ?

**Comment by fredbi** [comment]
Yes.  You just have to insert some type assertion switch on the return `error` to set it to the underlying error type and you are good to work with the content.

**Comment by deeptibaghel** [comment]
Many Thanks @fredbi
